WT-16386 Config parser reads beyond config length limit

ablekhman opened the pull request:
- Replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings (which may not be the case for configs) and reads beyond the valid buffer.

ablekhman committed:
WT-16386 Config parser reads beyond config length limit
- Replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings (which may not be the case for configs) and reads beyond the valid buffer.

pmacko86 commented on src/config/config.c:
```diff
@@ -8,6 +8,84 @@
+    const int cutlim = maxval % 10;
+
+    uint64_t acc = 0;
+    int flag = 0;
```
Maybe name this variable something else? Could this be a boolean variable, such as `error` or `failed`?

ablekhman replied:
This flag indicates several states:
- `0`: parsing unsuccessful, not a number
- `-1`: overflow  or underflow; `ERANGE` is set
- `1`: parsing successful

pmacko86 commented on src/config/config.c:
```diff
@@ -8,6 +8,84 @@
+    if (flag == -1) {
+        value = neg ? INT64_MIN : INT64_MAX;
+        errno = ERANGE;
+    } else {
```
Nit: No `{` and `}` around a single-statement body.

pmacko86 commented on src/config/config.c:
```diff
@@ -8,6 +8,84 @@
+        } else
+            break;
+
+        if (flag < 0 || acc > cutoff || (acc == cutoff && c > cutlim)) {
```
Nit: Here too – no `{` and `}` around a single-statement body.

ablekhman replied:
There is a multiline comment bellow. Omitting the bracers makes the code look quite strange.

pmacko86 commented on src/config/config.c:
```diff
@@ -8,6 +8,84 @@
+        value = neg ? -(int64_t)acc : (int64_t)acc;
+    }
+
+    if (endptr != NULL)
```
Please explain this in the function body comment above, especially about how this is set if parsing failed. Is this also how `strtoll` sets it on overflow?

ablekhman replied:
Yes, the intention is to make a replacement for `strtoll` so the code can simply call this function with the same side-effects, etc.

pmacko86 commented on test/catch2/misc_tests/test_config.cpp:
```diff
@@ -0,0 +1,44 @@
+    WT_SESSION_IMPL *session = conn.create_session();
+    WT_CONFIG cfg{};
+
+    SECTION("Limited length")
```
Please add more test cases, at least:
* Return 0 for no conversion.
* Boundary conditions, i.e., the smallest and the largest supported integers.
* Numbers out of the supported range, e.g., being just 1 out of the range, and being more one digit longer.
* Hopefully at least a few random numbers that should parse cleanly, both positive and negative—could be randomized.

To make this simpler, you could add the additional test by exporting a `_ut_` version of the parsing function for unit testing. In this way, you could perhaps more easily test error handling.

pmacko86 commented on src/config/config.c:
```diff
@@ -8,6 +8,84 @@
+ *      - If successful, an integer value corresponding to the contents of str is returned.
+ *      - If the converted value falls out of range, a range error occurs (setting errno to ERANGE)
+ *          and INT64_MAX or INT64_MIN is returned.
+ *      - If no conversion can be performed, 0 is returned.
```
Should this set `errno` to `EINVAL` if there are no characters to parse, to make this behave more like `strtoll`?

ablekhman replied:
I tried to make this function as close to `strtoll` as possible, except for the predefined base of 10. The intention is to have it as a drop-in replacement for `strtoll`.

pmacko86 reviewed (commented):
Almost LGTM. If we don't have a better standard library function to do this, this makes a lot of sense, and it's good to see this fixed.

I left a few comments about the semantics and a few minor comments about how to improve the code, but most importantly, please add more thorough unit tests for this.

ablekhman committed:
WT-16386 Config parser reads beyond config length limit
- Replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings (which may not be the case for configs) and reads beyond the valid buffer.
- Add unit-tests for the replacement function

ablekhman committed:
WT-16386 Config parser reads beyond config length limit
- Replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings (which may not be the case for configs) and reads beyond the valid buffer.
- Add unit-tests for the replacement function

pmacko86 commented on src/config/config.c:
```diff
@@ -8,6 +8,94 @@
+        value = neg ? INT64_MIN : INT64_MAX;
+        errno = ERANGE;
+    } else
+        value = neg ? -(int64_t)acc : (int64_t)acc;
```
The case in which `acc = abs(INT64_MIN) = 9223372036854775808` is interesting, as this value is larger than `INT64_MAX`, so `(int64_t)acc` would result in an overflow. I just tried this, and apparently `(int64_t)acc = -9223372036854775808`, but we also get `-(int64_t)acc = -9223372036854775808`, so it works, although my head hurts from trying to understand why. In any case, as this is already handled by a test, it should work.

ablekhman replied:
It works because the check above prevents `acc` from overflowing. Let's take `char` for example:
- max(char) = 127
- min(char) = -128
- max(unsigned char) = 255; this is our `acc`. Without negative sign, it can hold max or min value for signed char: 127 or 128.

The limits will be:
- `cutoff = 12`
- `cutlim = 7` - for positive numbers
- `cutlim = 8` - for negative numbers

The check `acc > cutoff || (acc == cutoff && c > cutlim)` will stop `acc` from overflowing respective `MAX_CHAR` or `MIN_CHAR`. The `acc > cutoff` part is important when we deal with bigger than allowed numbers, e.g. `s = "142"`. When `acc = 14` and there is another number ready to be added, then it's for sure an overflow/underflow.
Therefore, cast to signed type is safe.

pmacko86 commented on test/catch2/misc_tests/test_config.cpp:
```diff
@@ -0,0 +1,136 @@
+ * Unit tests for config-related functions.
+ */
+
+TEST_CASE("Parse integer", "[config][parse_int]")
```
Very nice test additions! To be extra, extra safe, you could also add a quick check for `INT64_MAX - 1` and `INT64_MIN + 1`, but it's probably ok.

ablekhman replied:
Good point. Added another test.

pmacko86 reviewed (approved):
LGTM, very nice cleanup and testing!

ablekhman committed:
WT-16386 Config parser reads beyond config length limit
- Replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings (which may not be the case for configs) and reads beyond the valid buffer.
- Add unit-tests for the replacement function
